src/legacy/core_plugins/data/public/expressions/expressions_service.ts

```diff
@@ -17,107 +17,51 @@
-// the interpreter plugin itself once they are ready
-import { Registry } from '@kbn/interpreter/common';
-import { Adapters } from 'src/plugins/inspector/public';
-import { Filter } from '@kbn/es-query';
-import { TimeRange } from 'src/plugins/data/public';
+import { setInspector, setInterpreter } from './services';
+import { execute } from './lib/execute';
+import { loader } from './lib/loader';
+import { render } from './lib/render';
 import { createRenderer } from './expression_renderer';
-import { createRunFn } from './expression_runner';
-import { Query } from '../query';
 
-export interface InitialContextObject {
-  timeRange?: TimeRange;
-  filters?: Filter[];
-  query?: Query;
-}
-
-export type getInitialContextFunction = () => InitialContextObject;
-
-export interface Handlers {
-  getInitialContext: getInitialContextFunction;
-  inspectorAdapters?: Adapters;
-  abortSignal?: AbortSignal;
-}
-
-type Context = object;
-export interface Result {
-  type: string;
-  as?: string;
-  value?: unknown;
-  error?: unknown;
-}
-
-interface RenderHandlers {
-  done: () => void;
-  onDestroy: (fn: () => void) => void;
-}
-
-export interface RenderFunction {
-  name: string;
-  displayName: string;
-  help: string;
-  validate: () => void;
-  reuseDomNode: boolean;
-  render: (domNode: Element, data: unknown, handlers: RenderHandlers) => void;
-}
-
-export type RenderFunctionsRegistry = Registry<unknown, RenderFunction>;
+import { Start as IInspector } from '../../../../../plugins/inspector/public';
 
-export interface Interpreter {
-  interpretAst(ast: Ast, context: Context, handlers: Handlers): Promise<Result>;
+export interface ExpressionsServiceStartDependencies {
+  inspector: IInspector;
 }
-
-type InterpreterGetter = () => Promise<{ interpreter: Interpreter }>;
-
-export interface ExpressionsServiceDependencies {
-  interpreter: {
-    renderersRegistry: RenderFunctionsRegistry;
-    getInterpreter: InterpreterGetter;
-  };
-}
-
 /**
  * Expressions Service
  * @internal
  */
 export class ExpressionsService {
-  public setup({
-    interpreter: { renderersRegistry, getInterpreter },
-  }: ExpressionsServiceDependencies) {
-    const run = createRunFn(
-      renderersRegistry,
-      getInterpreter().then(({ interpreter }) => interpreter)
-    );
+  public setup() {
+    // eslint-disable-next-line
+    const { getInterpreter } = require('../../../interpreter/public/interpreter');
+    getInterpreter()
+      .then(setInterpreter)
+      .catch((e: Error) => {
+        throw new Error('interpreter is not initialized');
+      });
+
+    return {
+      registerType: npSetup.plugins.data.expressions.registerType,
+      registerFunction: npSetup.plugins.data.expressions.registerFunction,
+      registerRenderer: npSetup.plugins.data.expressions.registerRenderer,
+    };
+  }
+
+  public start({ inspector }: ExpressionsServiceStartDependencies) {
+    const ExpressionRenderer = createRenderer(loader);
```

Comment:
Can we expose the renderer in the setup phase? Lens currently requires the expression renderer in the embeddable factory which should register it's stuff in the setup phase.

Reply:
yeah, also different plugins (will) register renderers, so in your setup lifecycle there is no guarantee that the functions or renderers exist in the registry yet.